clang/lib/Driver/Driver.cpp

```diff
@@ -1287,6 +1289,87 @@ void Driver::CreateOffloadingDeviceToolChains(Compilation &C,
         Diag(clang::diag::warn_drv_empty_joined_argument)
             << SYCLTargetsValues->getAsString(C.getInputArgs());
     }
+  }
+  // If the user specified --offload-arch, deduce the offloading
+  // target triple(s) from the set of architecture(s).
+  // Create a toolchain for each valid triple.
+  else if (HasValidSYCLRuntime &&
+           C.getInputArgs().hasArg(options::OPT_offload_arch_EQ) && !IsHIP &&
+           !IsCuda) {
+    const ToolChain *HostTC = C.getSingleOffloadToolChain<Action::OFK_Host>();
+    auto AMDTriple = getHIPOffloadTargetTriple(*this, C.getInputArgs());
+    auto NVPTXTriple = getNVIDIAOffloadTargetTriple(*this, C.getInputArgs(),
+                                                    HostTC->getTriple());
+
+    // Attempt to deduce the offloading triple from the set of architectures.
+    // We need to temporarily create these toolchains so that we can access
+    // tools for inferring architectures.
+    llvm::DenseSet<StringRef> Archs;
+    if (NVPTXTriple) {
+      auto TempTC = std::make_unique<toolchains::CudaToolChain>(
+          *this, *NVPTXTriple, *HostTC, C.getInputArgs(), Action::OFK_None);
+      for (StringRef Arch :
+           getOffloadArchs(C, C.getArgs(), Action::OFK_SYCL, &*TempTC, true))
+        Archs.insert(Arch);
+    }
+    if (AMDTriple) {
+      auto TempTC = std::make_unique<toolchains::AMDGPUOpenMPToolChain>(
+          *this, *AMDTriple, *HostTC, C.getInputArgs());
+      for (StringRef Arch :
+           getOffloadArchs(C, C.getArgs(), Action::OFK_SYCL, &*TempTC, true))
+        Archs.insert(Arch);
+    }
+    if (!AMDTriple && !NVPTXTriple) {
+      for (StringRef Arch :
+           getOffloadArchs(C, C.getArgs(), Action::OFK_SYCL, nullptr, true))
+        Archs.insert(Arch);
+    }
+    for (StringRef Arch : Archs) {
+      if (NVPTXTriple && IsNVIDIAOffloadArch(StringToOffloadArch(
+                             getProcessorFromTargetID(*NVPTXTriple, Arch)))) {
+        DerivedArchs[NVPTXTriple->getTriple()].insert(Arch);
+      } else if (AMDTriple &&
+                 IsAMDOffloadArch(StringToOffloadArch(
+                     getProcessorFromTargetID(*AMDTriple, Arch)))) {
+        DerivedArchs[AMDTriple->getTriple()].insert(Arch);
+      } else if (IsIntelCPUOffloadArch(StringToOffloadArchIntel(Arch))) {
+        DerivedArchs["spir64_x86_64"].insert(Arch);
+      } else if (IsIntelGPUOffloadArch(StringToOffloadArchIntel(Arch))) {
+        DerivedArchs["spir64_gen"].insert(Arch);
+      } else {
+        Diag(clang::diag::err_drv_invalid_sycl_target) << Arch;
+        return;
+      }
```

Comment:
```suggestion
      if (NVPTXTriple && IsNVIDIAOffloadArch(StringToOffloadArch(
                             getProcessorFromTargetID(*NVPTXTriple, Arch)))) {
        DerivedArchs[NVPTXTriple->getTriple()].insert(Arch);
        continue;
      } 
      if (AMDTriple &&
                 IsAMDOffloadArch(StringToOffloadArch(
                     getProcessorFromTargetID(*AMDTriple, Arch)))) {
        DerivedArchs[AMDTriple->getTriple()].insert(Arch);
        continue;
      }
      if (IsIntelCPUOffloadArch(StringToOffloadArchIntel(Arch))) {
        DerivedArchs["spir64_x86_64"].insert(Arch);
        continue;
      }
      if (IsIntelGPUOffloadArch(StringToOffloadArchIntel(Arch))) {
        DerivedArchs["spir64_gen"].insert(Arch);
        continue;
      }
      Diag(clang::diag::err_drv_invalid_sycl_target) << Arch;
      return;
```
Something like this.  Just a check then continue.  If none of these are satisfied, we error.  The reason why I suggested it is because the entire for loop is just if/else and there is no additional items to work against before the loop starts again.